__USC ID: 9907399097  ; 
Name: Subhiksha Rani__

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.svm import SVC, LinearSVC
from sklearn.metrics import hamming_loss, accuracy_score, silhouette_samples, silhouette_score
from sklearn.preprocessing import StandardScaler
from imblearn.over_sampling import SMOTE
from sklearn.cluster import KMeans
from collections import Counter
from scipy.spatial import distance

<br>__1. (a) Downloading the dataset & splitting the 70% data as training set & 30% data as test set.__

In [2]:
data = pd.read_csv("Frogs_MFCCs.csv")
training_data, test_data = train_test_split(data,test_size=0.30,random_state=42)


<br>__(b) i. Exact match and hamming score/ loss methods for evaluating multi-label classification.__

<br>__ii. Train a SVM for each of the labels, using Gaussian kernels and one versus all classifiers.__

__Finding best C & Gamma using cross-validation for Family label with raw attributes.__

In [3]:
params = {'C': [1, 10, 100, 1000], 'gamma': [0.001, 0.01, 0.1, 1], 'kernel': ['rbf']}
clf = GridSearchCV(cv=10, estimator=SVC(), param_grid=params, n_jobs=-1)
clf.fit(training_data.iloc[:,0:22], training_data.iloc[:,22:23].values.ravel())   
print('Best C:',clf.best_estimator_.C)
print('Best Gamma:',clf.best_estimator_.gamma)

Best C: 100
Best Gamma: 1


__Finding the Hamming Loss & Exact match ratio for Family label with raw attributes.__

In [4]:
clf1 = SVC(kernel="rbf", gamma=1, C=100)
clf1.fit(training_data.iloc[:,0:22], training_data.iloc[:,22:23].values.ravel())
y_pred1 = clf1.predict(test_data.iloc[:,0:22])
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,22:23].values.ravel(), y_pred1))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,22:23].values.ravel(), y_pred1))

Hamming Loss:  0.006484483557202408
Exact Match ratio:  0.9935155164427976


__Finding best C & Gamma using cross-validation for Genus label with raw attributes.__

In [5]:
clf2 = GridSearchCV(cv=10, estimator=SVC(), param_grid=params, n_jobs=-1)
clf2.fit(training_data.iloc[:,0:22], training_data.iloc[:,23:24].values.ravel())   
print('Best C:',clf2.best_estimator_.C)
print('Best Gamma:',clf2.best_estimator_.gamma)

Best C: 100
Best Gamma: 1


__Finding the Hamming Loss & Exact match ratio for Genus label with raw attributes.__

In [6]:
clf3 = SVC(kernel="rbf", gamma=1, C=100)
clf3.fit(training_data.iloc[:,0:22], training_data.iloc[:,23:24].values.ravel())
y_pred2 = clf3.predict(test_data.iloc[:,0:22])
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,23:24].values.ravel(), y_pred2))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,23:24].values.ravel(), y_pred2))

Hamming Loss:  0.007874015748031496
Exact Match ratio:  0.9921259842519685


__Finding best C & Gamma using cross-validation for Species label with raw attributes.__

In [7]:
clf4 = GridSearchCV(cv=10, estimator=SVC(), param_grid=params, n_jobs=-1)
clf4.fit(training_data.iloc[:,0:22], training_data.iloc[:,24:25].values.ravel())   
print('Best C:',clf4.best_estimator_.C)
print('Best Gamma:',clf4.best_estimator_.gamma)

Best C: 10
Best Gamma: 1


__Finding the Hamming Loss & Exact match ratio for Species label with raw attributes.__

In [8]:
clf5 = SVC(kernel="rbf", gamma=1, C=10)
clf5.fit(training_data.iloc[:,0:22], training_data.iloc[:,24:25].values.ravel())
y_pred3 = clf5.predict(test_data.iloc[:,0:22])
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,24:25].values.ravel(), y_pred3))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,24:25].values.ravel(), y_pred3))

Hamming Loss:  0.010653080129689671
Exact Match ratio:  0.9893469198703103


__Standardizing the features.__

In [9]:
scaler = StandardScaler() 
scaler.fit(training_data.iloc[:,0:22])
X_train_std = scaler.transform(training_data.iloc[:,0:22])  
X_test_std = scaler.transform(test_data.iloc[:,0:22])

__Finding best C & Gamma using cross-validation for Family label with standardized attributes.__

In [10]:
clf6 = GridSearchCV(cv=10, estimator=SVC(), param_grid=params, n_jobs=-1)
clf6.fit(X_train_std, training_data.iloc[:,22:23].values.ravel())   
print('Best C:',clf6.best_estimator_.C)
print('Best Gamma:',clf6.best_estimator_.gamma)

Best C: 10
Best Gamma: 0.1


__Finding the Hamming Loss & Exact match ratio for Family label with standardized attributes.__

In [11]:
clf7 = SVC(kernel="rbf", gamma=0.1, C=10)
clf7.fit(X_train_std, training_data.iloc[:,22:23].values.ravel())
y_pred4 = clf7.predict(X_test_std)
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,22:23].values.ravel(), y_pred4))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,22:23].values.ravel(), y_pred4))

Hamming Loss:  0.007410838351088467
Exact Match ratio:  0.9925891616489115


__Finding best C & Gamma using cross-validation for Genus label with standardized attributes.__

In [12]:
clf8 = GridSearchCV(cv=10, estimator=SVC(), param_grid=params, n_jobs=-1)
clf8.fit(X_train_std, training_data.iloc[:,23:24].values.ravel())   
print('Best C:',clf8.best_estimator_.C)
print('Best Gamma:',clf8.best_estimator_.gamma)

Best C: 10
Best Gamma: 0.1


__Finding the Hamming Loss & Exact match ratio for Genus label with standardized attributes.__

In [13]:
clf9 = SVC(kernel="rbf", gamma=0.1, C=10)
clf9.fit(X_train_std, training_data.iloc[:,23:24].values.ravel())
y_pred5 = clf9.predict(X_test_std)
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,23:24].values.ravel(), y_pred5))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,23:24].values.ravel(), y_pred5))

Hamming Loss:  0.012042612320518759
Exact Match ratio:  0.9879573876794813


__Finding best C & Gamma using cross-validation for Species label with standardized attributes.__

In [14]:
claf = GridSearchCV(cv=10, estimator=SVC(), param_grid=params, n_jobs=-1)
claf.fit(X_train_std, training_data.iloc[:,24:25].values.ravel())   
print('Best C:',claf.best_estimator_.C)
print('Best Gamma:',claf.best_estimator_.gamma)

Best C: 1000
Best Gamma: 0.01


__Finding the Hamming Loss & Exact match ratio for Species label with standardized attributes.__

In [15]:
claf1 = SVC(kernel="rbf", gamma=0.01, C=1000)
claf1.fit(X_train_std, training_data.iloc[:,24:25].values.ravel())
y_pred6 = claf1.predict(X_test_std)
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,24:25].values.ravel(), y_pred6))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,24:25].values.ravel(), y_pred6))

Hamming Loss:  0.010653080129689671
Exact Match ratio:  0.9893469198703103


<br>__iii. L1-penalized SVMs.__

__Finding best C using cross-validation for Family label.__

In [16]:
params_l1 = {'C': [1, 10, 100, 1000]}
claf2 = GridSearchCV(cv=10, estimator=LinearSVC(penalty='l1',random_state=0,loss='squared_hinge',dual=False,max_iter=10000), param_grid=params_l1, n_jobs=-1)
claf2.fit(X_train_std, training_data.iloc[:,22:23].values.ravel())   
print('Best C:',claf2.best_estimator_.C)

Best C: 1


__Finding the Hamming Loss & Exact match ratio for Family label.__

In [17]:
claf3 = LinearSVC(penalty='l1', random_state=0, C=1, loss='squared_hinge', dual=False, max_iter=10000)
claf3.fit(X_train_std, training_data.iloc[:,22:23].values.ravel())
y_pred7 = claf3.predict(X_test_std)
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,22:23].values.ravel(), y_pred7))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,22:23].values.ravel(), y_pred7))

Hamming Loss:  0.07132931912922649
Exact Match ratio:  0.9286706808707735


__Finding best C using cross-validation for Genus label.__

In [18]:
claf4 = GridSearchCV(cv=10, estimator=LinearSVC(penalty='l1',random_state=0,loss='squared_hinge',dual=False,max_iter=10000), param_grid=params_l1, n_jobs=-1)
claf4.fit(X_train_std, training_data.iloc[:,23:24].values.ravel())   
print('Best C:',claf4.best_estimator_.C)

Best C: 10


__Finding the Hamming Loss & Exact match ratio for Genus label.__

In [19]:
claf5 = LinearSVC(penalty='l1', random_state=0, C=10, loss='squared_hinge', dual=False, max_iter=10000)
claf5.fit(X_train_std, training_data.iloc[:,23:24].values.ravel())
y_pred8 = claf5.predict(X_test_std)
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,23:24].values.ravel(), y_pred8))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,23:24].values.ravel(), y_pred8))

Hamming Loss:  0.058360352014821676
Exact Match ratio:  0.9416396479851783


__Finding best C using cross-validation for Species label.__

In [20]:
claf6 = GridSearchCV(cv=10, estimator=LinearSVC(penalty='l1',random_state=0,loss='squared_hinge',dual=False,max_iter=10000), param_grid=params_l1, n_jobs=-1)
claf6.fit(X_train_std, training_data.iloc[:,24:25].values.ravel())   
print('Best C:',claf6.best_estimator_.C)

Best C: 1


__Finding the Hamming Loss & Exact match ratio for Species label.__

In [21]:
claf7 = LinearSVC(penalty='l1', random_state=0, C=1, loss='squared_hinge', dual=False, max_iter=10000)
claf7.fit(X_train_std, training_data.iloc[:,24:25].values.ravel())
y_pred9 = claf7.predict(X_test_std)
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,24:25].values.ravel(), y_pred9))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,24:25].values.ravel(), y_pred9))

Hamming Loss:  0.04075961093098657
Exact Match ratio:  0.9592403890690134


<br>__iv. SMOTE to deal with class imbalance.__

__Oversampling minority class using SMOTE for family label.__

In [22]:
sm = SMOTE(random_state=2)
X_train_sm, y_train_sm = sm.fit_sample(X_train_std, training_data.iloc[:,22:23].values.ravel())

__Finding best C using cross-validation for Family label.__

In [23]:
params_l1 = {'C': [1, 10, 100, 1000]}
claf8 = GridSearchCV(cv=10, estimator=LinearSVC(penalty='l1', random_state=0, loss='squared_hinge', dual=False, max_iter=10000), param_grid=params_l1, n_jobs=-1)
claf8.fit(X_train_sm, y_train_sm)   
print('Best C:',claf8.best_estimator_.C)

Best C: 10


__Finding the Hamming Loss & Exact match ratio for Family label.__

In [24]:
claf9 = LinearSVC(penalty='l1', random_state=0, C=10, loss='squared_hinge', dual=False, max_iter=10000)
claf9.fit(X_train_sm, y_train_sm)
pred_y = claf9.predict(X_test_std)
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,22:23].values.ravel(), pred_y))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,22:23].values.ravel(), pred_y))

Hamming Loss:  0.0921723019916628
Exact Match ratio:  0.9078276980083372


__Oversampling minority class using SMOTE for Genus label.__

In [25]:
sm = SMOTE(random_state=2)
X_train_sm1, y_train_sm1 = sm.fit_sample(X_train_std, training_data.iloc[:,23:24].values.ravel())

__Finding best C using cross-validation for Genus label.__

In [26]:
params_l1 = {'C': [1, 10, 100, 1000]}
cv1 = GridSearchCV(cv=10, estimator=LinearSVC(penalty='l1', random_state=0, loss='squared_hinge', dual=False, max_iter=10000), param_grid=params_l1, n_jobs=-1)
cv1.fit(X_train_sm1, y_train_sm1)   
print('Best C:',cv1.best_estimator_.C)

Best C: 10


__Finding the Hamming Loss & Exact match ratio for Genus label.__

In [27]:
svc1 = LinearSVC(penalty='l1', random_state=0, C=10, loss='squared_hinge', dual=False, max_iter=10000)
svc1.fit(X_train_sm1, y_train_sm1)
pred_y1 = svc1.predict(X_test_std)
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,23:24].values.ravel(), pred_y1))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,23:24].values.ravel(), pred_y1))

Hamming Loss:  0.0968040759610931
Exact Match ratio:  0.9031959240389069


__Oversampling minority class using SMOTE for Species label.__

In [28]:
sm = SMOTE(random_state=2)
X_train_sm2, y_train_sm2 = sm.fit_sample(X_train_std, training_data.iloc[:,24:25].values.ravel())

__Finding best C using cross-validation for Species label.__

In [29]:
cv2 = GridSearchCV(cv=10, estimator=LinearSVC(penalty='l1',random_state=0,loss='squared_hinge',dual=False,max_iter=10000), param_grid=params_l1, n_jobs=-1)
cv2.fit(X_train_sm2, y_train_sm2)   
print('Best C:',cv2.best_estimator_.C)

Best C: 100


__Finding the Hamming Loss & Exact match ratio for Species label.__

In [30]:
svc2 = LinearSVC(penalty='l1', random_state=0, C=100, loss='squared_hinge', dual=False, max_iter=10000)
svc2.fit(X_train_sm2, y_train_sm2)
pred_y2 = svc2.predict(X_test_std)
print("Hamming Loss: ", hamming_loss(test_data.iloc[:,24:25].values.ravel(), pred_y2))
print("Exact Match ratio: ", accuracy_score(test_data.iloc[:,24:25].values.ravel(), pred_y2))

Hamming Loss:  0.042612320518758684
Exact Match ratio:  0.9573876794812413


<br><br>__2. K-Means Clustering on a Multi-Class and Multi-Label Data Set.__

__(a) K-means clustering__

In [31]:
data_set = data

data_set = data_set.replace('Bufonidae',0)
data_set = data_set.replace('Dendrobatidae',1)
data_set = data_set.replace('Hylidae',2)
data_set = data_set.replace('Leptodactylidae',3)

data_set = data_set.replace('Adenomera',0)
data_set = data_set.replace('Ameerega',1)
data_set = data_set.replace('Dendropsophus',2)
data_set = data_set.replace('Hypsiboas',3)
data_set = data_set.replace('Leptodactylus',4)
data_set = data_set.replace('Osteocephalus',5)
data_set = data_set.replace('Rhinella',6)
data_set = data_set.replace('Scinax',7)

data_set = data_set.replace('AdenomeraAndre',0)
data_set = data_set.replace('AdenomeraHylaedactylus',1)
data_set = data_set.replace('Ameeregatrivittata',2)
data_set = data_set.replace('HylaMinuta',3)
data_set = data_set.replace('HypsiboasCinerascens',4)
data_set = data_set.replace('HypsiboasCordobae',5)
data_set = data_set.replace('LeptodactylusFuscus',6)
data_set = data_set.replace('OsteocephalusOophagus',7)
data_set = data_set.replace('Rhinellagranulosa',8)
data_set = data_set.replace('ScinaxRuber',9)

In [32]:
clusters = []
sil_avg = []
for n_clusters in range(2,51):
    clusterer = KMeans(n_clusters=n_clusters, random_state=10)
    cluster_labels = clusterer.fit_predict(data_set.iloc[:,0:22])
    silhouette_avg = silhouette_score(data_set.iloc[:,0:22], cluster_labels, metric='euclidean')
    print("For n_clusters =", n_clusters, "The average silhouette_score is :", silhouette_avg)
    clusters.append(n_clusters)
    sil_avg.append(silhouette_avg)
scores = pd.DataFrame()
scores.insert(0,'n_clusters',clusters)  
scores.insert(1,'Silhouette score',sil_avg)

For n_clusters = 2 The average silhouette_score is : 0.3486778410277152
For n_clusters = 3 The average silhouette_score is : 0.36768245219926315
For n_clusters = 4 The average silhouette_score is : 0.3787509343305295
For n_clusters = 5 The average silhouette_score is : 0.3714828057802203
For n_clusters = 6 The average silhouette_score is : 0.2639575223654826
For n_clusters = 7 The average silhouette_score is : 0.26063587785642767
For n_clusters = 8 The average silhouette_score is : 0.25476282597833133
For n_clusters = 9 The average silhouette_score is : 0.2762425838810639
For n_clusters = 10 The average silhouette_score is : 0.2624669280635291
For n_clusters = 11 The average silhouette_score is : 0.2714493919668881
For n_clusters = 12 The average silhouette_score is : 0.27114249474339275
For n_clusters = 13 The average silhouette_score is : 0.2746296729972097
For n_clusters = 14 The average silhouette_score is : 0.28295590739731163
For n_clusters = 15 The average silhouette_score is : 

In [33]:
print('Best Silhouette score: ', scores['Silhouette score'].max(),' is obtained for n_cluster= ',scores.loc[scores['Silhouette score']==scores['Silhouette score'].max(), 'n_clusters'].iloc[0])

Best Silhouette score:  0.3787509343305295  is obtained for n_cluster=  4


__(b) In each cluster, determine which family is the majority by reading the true labels. Repeat for genus and species.__

In [34]:
clusterer = KMeans(n_clusters=4, random_state=10)
cluster_labels = clusterer.fit_predict(data_set.iloc[:,0:22])

In [35]:
cluster0_family = []
cluster0_genus = []
cluster0_species = []
cluster1_family = []
cluster1_genus = []
cluster1_species = []
cluster2_family = []
cluster2_genus = []
cluster2_species = []
cluster3_family = []
cluster3_genus = []
cluster3_species = []
cluster0_pred = []
cluster1_pred = []
cluster2_pred = []
cluster3_pred = []
for i in range(0,7195):
        if cluster_labels[i] == 0:
            cluster0_family.append(data_set['Family'][i])
            cluster0_genus.append(data_set['Genus'][i])
            cluster0_species.append(data_set['Species'][i])
            cluster0_pred.append(cluster_labels[i])
        elif cluster_labels[i] == 1:
            cluster1_family.append(data_set['Family'][i])
            cluster1_genus.append(data_set['Genus'][i])
            cluster1_species.append(data_set['Species'][i])
            cluster1_pred.append(cluster_labels[i])
        elif cluster_labels[i] == 2:
            cluster2_family.append(data_set['Family'][i])
            cluster2_genus.append(data_set['Genus'][i])
            cluster2_species.append(data_set['Species'][i])
            cluster2_pred.append(cluster_labels[i])
        elif cluster_labels[i] == 3:
            cluster3_family.append(data_set['Family'][i])
            cluster3_genus.append(data_set['Genus'][i])
            cluster3_species.append(data_set['Species'][i])
            cluster3_pred.append(cluster_labels[i])

In [36]:
family0 = Counter(cluster0_family)
print(family0.most_common(1))

[(3, 3467)]


__The most common Family in Cluster 0 is Leptodactylidae with a count of 3467.__

In [37]:
genus0 = Counter(cluster0_genus)
print(genus0.most_common(1))

[(0, 3466)]


__The most common Genus in Cluster 0 is Adenomera with a count of 3466.__

In [38]:
species0 = Counter(cluster0_species)
print(species0.most_common(1))

[(1, 3466)]


__The most common Species in Cluster 0 is AdenomeraHylaedactylus with a count of 3466.__

In [39]:
family1 = Counter(cluster1_family)
print(family1.most_common(1))

[(2, 1245)]


__The most common Family in Cluster 1 is Hylidae with a count of 1245.__

In [40]:
genus1 = Counter(cluster1_genus)
print(genus1.most_common(1))

[(3, 1038)]


__The most common Genus in Cluster 1 is Hypsiboas with a count of 1038.__

In [41]:
species1 = Counter(cluster1_species)
print(species1.most_common(1))

[(5, 1018)]


__The most common Species in Cluster 1 is HypsiboasCordobae with a count of 1018.__

In [42]:
family2 = Counter(cluster2_family)
print(family2.most_common(1))

[(1, 500)]


__The most common Family in Cluster 2 is Dendrobatidae with a count of 500.__

In [43]:
genus2 = Counter(cluster2_genus)
print(genus2.most_common(1))

[(1, 500)]


__The most common Genus in Cluster 2 is Ameerega with a count of 500.__

In [44]:
species2 = Counter(cluster2_species)
print(species2.most_common(1))

[(2, 500)]


__The most common Species in Cluster 2 is Ameeregatrivittata with a count of 500.__

In [45]:
family3 = Counter(cluster3_family)
print(family3.most_common(1))

[(2, 590)]


__The most common Family in Cluster 3 is Hylidae with a count of 590.__

In [46]:
genus3 = Counter(cluster3_genus)
print(genus3.most_common(1))

[(3, 542)]


__The most common Genus in Cluster 3 is Hypsiboas with a count of 542.__

In [47]:
species3 = Counter(cluster3_species)
print(species3.most_common(1))

[(4, 452)]


__The most common Species in Cluster 3 is HypsiboasCinerascens with a count of 452.__

<br>__(c)Hamming Distance, Hamming Loss & Hamming Score for each cluster.__

In [48]:
hd0 = (distance.hamming(cluster0_family,cluster0_pred)+distance.hamming(cluster0_genus,cluster0_pred)+distance.hamming(cluster0_species,cluster0_pred))/3

d = len(cluster0_pred)

correct_fam0 = 0
wrong_fam0 = 0
for i in cluster0_family:
    if i == 3:
        correct_fam0 = correct_fam0+1
    else:
        wrong_fam0 = wrong_fam0+1

correct_gen0 = 0
wrong_gen0 = 0
for i in cluster0_genus:
    if i == 0:
        correct_gen0 = correct_gen0+1
    else:
        wrong_gen0 = wrong_gen0+1

correct_spe0 = 0
wrong_spe0 = 0
for i in cluster0_species:
    if i == 1:
        correct_spe0 = correct_spe0+1
    else:
        wrong_spe0 = wrong_spe0+1

hl0 = ((wrong_fam0/d)+(wrong_gen0/d)+(wrong_spe0/d))/3
hs0 = ((correct_fam0/(d+d))+(correct_gen0/(d+d))+(correct_spe0/(d+d)))/3

print("Hamming Distance for Cluster 0: ", hd0)
print("Hamming Loss for Cluster 0: ", hl0)
print("Hamming Score for Cluster 0: ", hs0)

Hamming Distance for Cluster 0:  0.6761958146487294
Hamming Loss for Cluster 0:  0.02849402092675635
Hamming Score for Cluster 0:  0.4857529895366219


In [49]:
hd1 = (distance.hamming(cluster1_family,cluster1_pred)+distance.hamming(cluster1_genus,cluster1_pred)+distance.hamming(cluster1_species,cluster1_pred))/3

d = len(cluster1_pred)
correct_fam1 = 0
wrong_fam1 = 0
for i in cluster1_family:
    if i == 2:
        correct_fam1 = correct_fam1+1
    else:
        wrong_fam1 = wrong_fam1+1

correct_gen1 = 0
wrong_gen1 = 0
for i in cluster1_genus:
    if i == 3:
        correct_gen1 = correct_gen1+1
    else:
        wrong_gen1 = wrong_gen1+1

correct_spe1 = 0
wrong_spe1 = 0
for i in cluster1_species:
    if i == 5:
        correct_spe1 = correct_spe1+1
    else:
        wrong_spe1 = wrong_spe1+1

hl1 = ((wrong_fam1/d)+(wrong_gen1/d)+(wrong_spe1/d))/3
hs1 = ((correct_fam1/(d+d))+(correct_gen1/(d+d))+(correct_spe1/(d+d)))/3

print("Hamming Distance for Cluster 1: ", hd1)
print("Hamming Loss for Cluster 1: ", hl1)
print("Hamming Score for Cluster 1: ", hs1)

Hamming Distance for Cluster 1:  0.9848637739656912
Hamming Loss for Cluster 1:  0.44483686511940795
Hamming Score for Cluster 1:  0.277581567440296


In [50]:
hd2 = (distance.hamming(cluster2_family,cluster2_pred)+distance.hamming(cluster2_genus,cluster2_pred)+distance.hamming(cluster2_species,cluster2_pred))/3

d = len(cluster2_pred)
correct_fam2 = 0
wrong_fam2 = 0
for i in cluster2_family:
    if i == 1:
        correct_fam2 = correct_fam2+1
    else:
        wrong_fam2 = wrong_fam2+1

correct_gen2 = 0
wrong_gen2 = 0
for i in cluster2_genus:
    if i == 1:
        correct_gen2 = correct_gen2+1
    else:
        wrong_gen2 = wrong_gen2+1

correct_spe2 = 0
wrong_spe2 = 0
for i in cluster2_species:
    if i == 2:
        correct_spe2 = correct_spe2+1
    else:
        wrong_spe2 = wrong_spe2+1

hl2 = ((wrong_fam2/d)+(wrong_gen2/d)+(wrong_spe2/d))/3
hs2 = ((correct_fam2/(d+d))+(correct_gen2/(d+d))+(correct_spe2/(d+d)))/3

print("Hamming Distance for Cluster 2: ", hd2)
print("Hamming Loss for Cluster 2: ", hl2)
print("Hamming Score for Cluster 2: ", hs2)

Hamming Distance for Cluster 2:  0.7032007759456839
Hamming Loss for Cluster 2:  0.5150339476236664
Hamming Score for Cluster 2:  0.24248302618816683


In [51]:
hd3 = (distance.hamming(cluster3_family,cluster3_pred)+distance.hamming(cluster3_genus,cluster3_pred)+distance.hamming(cluster3_species,cluster3_pred))/3

d = len(cluster3_pred)
correct_fam3 = 0
wrong_fam3 = 0
for i in cluster3_family:
    if i == 2:
        correct_fam3 = correct_fam3+1
    else:
        wrong_fam3 = wrong_fam3+1

correct_gen3 = 0
wrong_gen3 = 0
for i in cluster3_genus:
    if i == 3:
        correct_gen3 = correct_gen3+1
    else:
        wrong_gen3 = wrong_gen3+1

correct_spe3 = 0
wrong_spe3 = 0
for i in cluster3_species:
    if i == 4:
        correct_spe3 = correct_spe3+1
    else:
        wrong_spe3 = wrong_spe3+1

hl3 = ((wrong_fam3/d)+(wrong_gen3/d)+(wrong_spe3/d))/3
hs3 = ((correct_fam3/(d+d))+(correct_gen3/(d+d))+(correct_spe3/(d+d)))/3

print("Hamming Distance for Cluster 3: ", hd3)
print("Hamming Loss for Cluster 3: ", hl3)
print("Hamming Score for Cluster 3: ", hs3)

Hamming Distance for Cluster 3:  0.6938110749185668
Hamming Loss for Cluster 3:  0.14006514657980454
Hamming Score for Cluster 3:  0.4299674267100977
